waku/common/enr/builder.nim

```diff
@@ -0,0 +1,103 @@
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+
+import
+  std/options,
+  stew/results,
+  stew/shims/net,
+  eth/keys as eth_keys,
+  eth/p2p/discoveryv5/enr,
+  libp2p/crypto/crypto as libp2p_crypto
+
+
+## Builder
+
+type EnrBuilder* = object
+    seqNumber: uint64
+    privateKey: eth_keys.PrivateKey
+    fields: seq[FieldPair]
+
+
+proc init*(T: type EnrBuilder, key: eth_keys.PrivateKey, seqNum: uint64 = 1): EnrBuilder =
+  EnrBuilder(
+    seqNumber: seqNum,
+    privateKey: key,
+    fields: newSeq[FieldPair]()
+  )
+
+proc init*(T: type EnrBuilder, key: libp2p_crypto.PrivateKey, seqNum: uint64 = 1): EnrBuilder =
+  # TODO: Inconvenient runtime assertion. Move this assertion to compile time
+  if key.scheme != PKScheme.Secp256k1:
+    raise newException(Defect, "invalid private key scheme")
+
+  let
+    bytes = key.getRawBytes().expect("Private key is valid")
+    privateKey = eth_keys.PrivateKey.fromRaw(bytes).expect("Raw private key is of valid length")
+
+  EnrBuilder.init(key=privateKey, seqNum=seqNum)
+
+proc addFieldPair*(builder: var EnrBuilder, pair: FieldPair) =
+  builder.fields.add(pair)
+
+proc addFieldPair*[V](builder: var EnrBuilder, key: string, value: V) =
+  builder.addFieldPair(toFieldPair(key, value))
```

Comment:
we may want to add a check that `value` is not null here - this caused enrs to be misread by other implementations

Reply:
What do you mean by "is not `null`"? Could you provide a snippet illustrating the scenario where the `value` argument could be null?

Reply:
err, not directly related, but https://github.com/waku-org/nwaku/pull/1559

do you intend `addFieldPair` to only be used by the enr builder module?

Reply:
> err, not directly related, but #1559
> 
> do you intend `addFieldPair` to only be used by the enr builder module?

I see. This field should not be used outside the "extension modules". If used, it is the responsibility of the developer to perform the checks.

Regarding #1559, I plan to rework the multiaddress extension as part of the static sharding implementation. I will also add the multiaddress validation to the builder extension. Thanks for the heads-up 👍🏼

Reply:
Sounds good! Thanks